process/block/headerForBlock/headersForBlock.go

```diff
@@ -453,53 +459,71 @@ func (hfb *headersForBlock) computeExistingAndRequestMissingMetaHeaders(header d
 	}
 }
 
-func (hfb *headersForBlock) computeExistingAndRequestMissingShardHeaders(metaBlock data.MetaHeaderHandler) {
-	hfb.mutHdrsForBlock.Lock()
-	defer hfb.mutHdrsForBlock.Unlock()
-
-	for _, shardData := range metaBlock.GetShardInfoHandlers() {
-		if shardData.GetNonce() == hfb.genesisNonce {
-			lastCrossNotarizedHeaderForShard, found := hfb.lastNotarizedShardHeaders[shardData.GetShardID()]
-			if !found {
-				log.Warn("computeExistingAndRequestMissingShardHeaders.GetLastCrossNotarizedHeader could not find last notarized", "shard", shardData.GetShardID())
-				continue
-			}
-			if !bytes.Equal(lastCrossNotarizedHeaderForShard.GetHash(), shardData.GetHeaderHash()) {
-				log.Warn("genesis hash mismatch",
-					"last notarized hash", lastCrossNotarizedHeaderForShard.GetHash(),
-					"genesis nonce", hfb.genesisNonce,
-					"genesis hash", shardData.GetHeaderHash())
-			}
-			continue
+func (hfb *headersForBlock) requestMissingAndUpdateBasedOnCrossShardData(cd crossShardMetaData) {
+	if cd.GetNonce() == hfb.genesisNonce {
+		lastCrossNotarizedHeaderForShard, found := hfb.lastNotarizedShardHeaders[cd.GetShardID()]
+		if !found {
+			log.Warn("requestMissingAndUpdateBasedOnCrossShardData could not find last notarized", "shard", cd.GetShardID())
+			return
 		}
+		if !bytes.Equal(lastCrossNotarizedHeaderForShard.GetHash(), cd.GetHeaderHash()) {
+			log.Warn("genesis hash mismatch",
+				"last notarized hash", lastCrossNotarizedHeaderForShard.GetHash(),
+				"genesis nonce", hfb.genesisNonce,
+				"genesis hash", cd.GetHeaderHash())
+		}
+		return
+	}
 
-		hdr, err := process.GetShardHeaderFromPool(
-			shardData.GetHeaderHash(),
-			hfb.dataPool.Headers())
+	hdr, err := process.GetShardHeaderFromPool(
+		cd.GetHeaderHash(),
+		hfb.dataPool.Headers())
 
-		if err != nil {
-			hfb.missingHdrs++
-			hfb.hdrHashAndInfo[string(shardData.GetHeaderHash())] = newHeaderInfo(nil, true, false, false)
+	if err != nil {
+		hfb.missingHdrs++
+		hfb.hdrHashAndInfo[string(cd.GetHeaderHash())] = newHeaderInfo(nil, true, false, false)
 
-			go hfb.requestHandler.RequestShardHeader(shardData.GetShardID(), shardData.GetHeaderHash())
-			continue
-		}
+		go hfb.requestHandler.RequestShardHeader(cd.GetShardID(), cd.GetHeaderHash())
+		return
+	}
 
-		hfb.hdrHashAndInfo[string(shardData.GetHeaderHash())] = newHeaderInfo(hdr, true, false, false)
+	hfb.hdrHashAndInfo[string(cd.GetHeaderHash())] = newHeaderInfo(hdr, true, false, false)
```

Comment:
[comment]
Just an observation.
Here, after we find the request shard header in pool, we set `hasProof` by default to false and request it afterwards. Would it make sense to actually check the proof is missing(check in pool) before requesting it? Or is it guaranteed that the proof is missing if this function is called

Reply:
the check is done inside requestProofIfNeeded and request is done only if it is not present (it also updates the hasProof flag.